rcl/test/rcl/test_graph.cpp

```diff
@@ -0,0 +1,153 @@
+  rcl_client_options_t client_options = rcl_client_get_default_options();
+  ret = rcl_client_init(&client, this->node_ptr, ts, service_name, &client_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  auto client_exit = make_scope_exit([&client, this]() {
+    stop_memory_checking();
+    rcl_ret_t ret = rcl_client_fini(&client, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  });
+  // Check, knowing there is no service server (created by us at least).
+  bool is_available;
+  ret = rcl_service_server_is_available(this->node_ptr, &client, &is_available);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  ASSERT_FALSE(is_available);
+  // Create the service server.
+  rcl_service_t service = rcl_get_zero_initialized_service();
+  rcl_service_options_t service_options = rcl_service_get_default_options();
+  ret = rcl_service_init(&service, this->node_ptr, ts, service_name, &service_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  auto service_exit = make_scope_exit([&service, this]() {
+    stop_memory_checking();
+    rcl_ret_t ret = rcl_service_fini(&service, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  });
+  // Wait for the graph state to change, check, and try again for a period of time.
+  is_available = false;
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  ret = rcl_wait_set_init(&wait_set, 0, 1, 0, 0, 0, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  auto wait_set_exit = make_scope_exit([&wait_set, this]() {
+    stop_memory_checking();
+    rcl_ret_t ret = rcl_wait_set_fini(&wait_set);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  });
+  const rcl_guard_condition_t * graph_guard_condition =
+    rcl_node_get_graph_guard_condition(this->node_ptr);
+  ASSERT_NE(nullptr, graph_guard_condition) << rcl_get_error_string_safe();
+  auto start = std::chrono::steady_clock::now();
+  auto end = start + std::chrono::seconds(10);
+  while (std::chrono::steady_clock::now() < end) {
+    // We wait multiple times in case other graph changes are occurring simultaneously.
+    std::chrono::nanoseconds time_left = end - start;
+    std::chrono::nanoseconds time_to_sleep = time_left;
+    std::chrono::seconds min_sleep(1);
+    if (time_to_sleep > min_sleep) {
+      time_to_sleep = min_sleep;
+    }
+    ret = rcl_wait_set_clear_guard_conditions(&wait_set);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+    ret = rcl_wait_set_add_guard_condition(&wait_set, graph_guard_condition);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+    printf("waiting up to %lld nanoseconds for graph changes\n", time_to_sleep.count());
+    ret = rcl_wait(&wait_set, time_to_sleep.count());
+    if (ret == RCL_RET_TIMEOUT) {
+      printf("timeout!\n");
```

Comment:
Sorry, that was left over from debug. I removed it. There's an identical check one below.

Incidentally, I have a few more changes for this pr.